File: leetcode/string/substring_with_concatenation_of_all_words.go
package main

import "fmt"

func findSubstring(s string, words []string) []int {
	if len(words) == 0 || s == "" {
		return nil
	}
	lw, lws, ls := len(words[0]), len(words), len(s)
	totallen := lw * lws
	if ls < totallen {
		return nil
	}
	var rst []int
	ct := counter(words)
	for i := 0; i < ls-totallen+1; i++ {
		ctt, j := make(map[string]int), 0
		for j < lws {
			curr := s[i+j*lw : i+(j+1)*lw]
			if _, ok := ctt[curr]; ok {
				ctt[curr]++
			} else {
				ctt[curr] = 1
			}

			if ctt[curr] > ct[curr] {
				break
			}
			j++
		}
		if j == lws {
			rst = append(rst, i)
		}
	}
	return rst
}

func counter(words []string) map[string]int {
	m := make(map[string]int)
	for _, word := range words {
		if _, ok := m[word]; ok {
			m[word]++
		} else {
			m[word] = 1
		}
	}
	return m
}

func main() {
	words := []string{"foo", "bar"}
	fmt.Println(findSubstring("barfoothefoobarman", words))
}

File: leetcode/string/substring_with_concatenation_of_all_words.py
from typing import List
import collections

class Solution:
    def findSubString(self, s:str, words: List[str]) -> List[str]:
        if not words: return []
        lw, lws, ls = len(words[0]), len(words), len(s)
        totallen = lw*lws
        if ls < totallen: return []
        rst, ct = [], collections.Counter(words)
        for i in range(ls - totallen + 1):
            ctt, j = {}, 0
            while j < lws:
                curr = s[i + j*lw: i+(j+1)*lw]
                ctt[curr] = ctt.get(curr, 0) + 1
                if ctt[curr] > ct[curr]: break
                j += 1
            if j == lws: rst.append(i)
        return rst

if __name__ == "__main__":
    print(Solution().findSubString("barfoothefoobarman", ["foo", "bar"]))
